### PR TITLE
[HybridParallel]Add scatter-gather for pipeline

### PR DESCRIPTION
**Pull request by ForFishes**
### PR types

Performance optimization
### PR changes

Others
### Describe

[HybridParallel]Add scatter-gather for pipeline
相对于静态图的scatter-gather优化，这个是动态图的版本.
https://github.com/PaddlePaddle/Paddle/pull/34110
image

image

### CODE REVIEW

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py [hunk 1]
```diff
@@ -377,24 +382,53 @@ def _send_activations(self, cache_id):
...
             for output in outputs:
-                p2p.send(output, self.next_stage_id)
+                if self.is_pipe_partitioned:
+                    p2p.send_partial(
+                        output,
+                        self.next_stage_id,
+                        mp_degree=self.mp_degree,
+                        mp_rank=self.mp_rank)
+                else:
+                    p2p.send(output.detach(), self.next_stage_id)
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py:409]: detach是防止output后面被改写吗，send_partial为啥没加呢

**Review by wangxicoding — DISMISSED** [review]
LGTM

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py (hunk 1 above)
Comment (ForFishes) [reply python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py:409]: 应该需要添加。主要目的反正梯度回传。

**Review by zhiqiu — COMMENTED** [review]
LGTM for op_function_generator.cc

**Review by zhiqiu — APPROVED** [review]

**Pull request merged**